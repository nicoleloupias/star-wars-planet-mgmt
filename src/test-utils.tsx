import { ChakraProvider } from "@chakra-ui/react";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import React from "react";
import { Router, RouterProvider } from "react-router-dom";
import { theme } from "./theme";
const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <ChakraProvider theme={theme}>
      <Router location={history.location} navigator={history}>
        {children}
      </Router>
    </ChakraProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {}
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => <AllProviders history={history}>{children}</AllProviders>,
    ...options
  });

export { customRender as render };
