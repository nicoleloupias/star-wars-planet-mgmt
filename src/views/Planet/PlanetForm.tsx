import { Button, Flex, Stack } from "@chakra-ui/react";
import { InputController } from "../../components/InputController";
import { useFormContext } from "react-hook-form";
import { maxLength, required } from "../../helpers/validations";
import { useNavigate } from "react-router-dom";

export interface PlanetFormFieldValues {
  name: string;
  diameter: number;
  population: number;
  climate: string;
  terrain: string;
}

export interface PlanetFormProps {
  onSubmit: () => void;
  confirmText: string;
}

export const PlanetForm = ({ onSubmit, confirmText }: PlanetFormProps) => {
  const { control, formState } = useFormContext<PlanetFormFieldValues>();
  const navigate = useNavigate();

  return (
    <Stack as="form" w="full" spacing={5}>
      <InputController
        id="name"
        control={control}
        label="Name"
        rules={{
          required,
          maxLength: maxLength(100)
        }}
      />
      <InputController id="diameter" type="number" label="Diameter (km)" control={control} />
      <InputController id="population" type="number" label="Population" control={control} />
      <InputController
        id="climate"
        label="Climate"
        control={control}
        rules={{
          required
        }}
      />
      <InputController
        id="terrain"
        label="Terrain"
        control={control}
        rules={{
          required
        }}
      />
      <Flex gap={3} pt={4}>
        <Button
          variant="secondary"
          onClick={() => {
            navigate(-1);
          }}
        >
          Cancel
        </Button>
        <Button
          variant="primary"
          type="submit"
          onClick={(e) => {
            e.preventDefault();
            onSubmit();
          }}
          isDisabled={!formState.isValid}
        >
          {confirmText}
        </Button>
      </Flex>
    </Stack>
  );
};
