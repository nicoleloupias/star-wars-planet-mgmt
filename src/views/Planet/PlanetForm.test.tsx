import { describe, expect, test, vi } from "vitest";
import { render } from "../../test-utils";
import type { RenderResult } from "@testing-library/react";
import { act, fireEvent, waitFor } from "@testing-library/react";
import type { PlanetFormProps } from "./PlanetForm";
import { PlanetForm } from "./PlanetForm";
import { FormProvider, useForm } from "react-hook-form";

describe("<PlanetForm/>", () => {
  const defaultProps = {
    onSubmit: vi.fn(),
    confirmText: "Create"
  };

  const Component = (props: PlanetFormProps) => {
    const form = useForm();
    return (
      <FormProvider {...form}>
        <PlanetForm {...props} />
      </FormProvider>
    );
  };
  const factoryComponent = (props: PlanetFormProps = defaultProps) => {
    return render(<Component {...props} />);
  };

  const fillForm = (component: RenderResult) => {
    act(() => {
      fireEvent.change(component.getByRole("textbox", { name: "Name" }), { target: { value: "Alex" } });
      fireEvent.change(component.getByRole("textbox", { name: "Climate" }), { target: { value: "Hot" } });
      fireEvent.change(component.getByRole("spinbutton", { name: "Population" }), { target: { value: "1000" } });
      fireEvent.change(component.getByRole("textbox", { name: "Terrain" }), { target: { value: "Terrain" } });
      fireEvent.change(component.getByRole("spinbutton", { name: "Diameter (km)" }), { target: { value: "150" } });
    });
  };

  test("should render succesfully", () => {
    const component = factoryComponent();

    expect(component).toMatchSnapshot();
  });

  test("should render elements correctly", () => {
    const component = factoryComponent();

    expect(component.getByRole("textbox", { name: "Name" })).toBeInTheDocument();
    expect(component.getByRole("textbox", { name: "Climate" })).toBeInTheDocument();
    expect(component.getByRole("spinbutton", { name: "Population" })).toBeInTheDocument();
    expect(component.getByRole("textbox", { name: "Terrain" })).toBeInTheDocument();
    expect(component.getByRole("spinbutton", { name: "Diameter (km)" })).toBeInTheDocument();
  });

  test("confirm btn should be disabled until form is filled correctly", async () => {
    const component = factoryComponent();

    expect(component.getByRole("button", { name: /Create/ })).toBeDisabled();

    fillForm(component);

    await waitFor(() => {
      expect(component.getByRole("button", { name: /Create/ })).toBeEnabled();
    });
  });

  test("should fire onSubmit when confirm btn is clicked", async () => {
    const component = factoryComponent();

    const confirmBtn = component.getByRole("button", { name: /Create/ });

    fillForm(component);

    await waitFor(() => {
      expect(component.getByRole("button", { name: /Create/ })).toBeEnabled();
    });

    fireEvent.click(confirmBtn);

    expect(defaultProps.onSubmit).toHaveBeenCalled();
  });

  test("should have custom button text if confirmText is passed", async () => {
    const component = factoryComponent({ ...defaultProps, confirmText: "Edit" });

    expect(component.getByRole("button", { name: /Edit/ })).toBeInTheDocument();
  });
});
