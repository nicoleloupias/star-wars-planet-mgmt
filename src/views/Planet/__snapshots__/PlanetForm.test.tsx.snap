// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PlanetForm/> > should render succesfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div>
      <form
        class="chakra-stack css-g6wlko"
      >
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="name"
            id="field-:r0:-label"
          >
            Name
          </label>
          <input
            class="chakra-input css-qcorp"
            data-testid="name"
            id="name"
            name="name"
            placeholder="Name"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="diameter"
            id="field-:r1:-label"
          >
            Diameter
          </label>
          <input
            class="chakra-input css-qcorp"
            data-testid="diameter"
            id="diameter"
            name="diameter"
            placeholder="Diameter"
            type="number"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="population"
            id="field-:r2:-label"
          >
            Population
          </label>
          <input
            class="chakra-input css-qcorp"
            data-testid="population"
            id="population"
            name="population"
            placeholder="Population"
            type="number"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="climate"
            id="field-:r3:-label"
          >
            Climate
          </label>
          <input
            class="chakra-input css-qcorp"
            data-testid="climate"
            id="climate"
            name="climate"
            placeholder="Climate"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="terrain"
            id="field-:r4:-label"
          >
            Terrain
          </label>
          <input
            class="chakra-input css-qcorp"
            data-testid="terrain"
            id="terrain"
            name="terrain"
            placeholder="Terrain"
            value=""
          />
        </div>
        <div
          class="css-u8chpe"
        >
          <button
            class="chakra-button css-4g9tt9"
            type="button"
          >
            Cancel
          </button>
          <button
            class="chakra-button css-12euwm4"
            disabled=""
            type="submit"
          >
            Create
          </button>
        </div>
      </form>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <form
      class="chakra-stack css-g6wlko"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="name"
          id="field-:r0:-label"
        >
          Name
        </label>
        <input
          class="chakra-input css-qcorp"
          data-testid="name"
          id="name"
          name="name"
          placeholder="Name"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="diameter"
          id="field-:r1:-label"
        >
          Diameter
        </label>
        <input
          class="chakra-input css-qcorp"
          data-testid="diameter"
          id="diameter"
          name="diameter"
          placeholder="Diameter"
          type="number"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="population"
          id="field-:r2:-label"
        >
          Population
        </label>
        <input
          class="chakra-input css-qcorp"
          data-testid="population"
          id="population"
          name="population"
          placeholder="Population"
          type="number"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="climate"
          id="field-:r3:-label"
        >
          Climate
        </label>
        <input
          class="chakra-input css-qcorp"
          data-testid="climate"
          id="climate"
          name="climate"
          placeholder="Climate"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="terrain"
          id="field-:r4:-label"
        >
          Terrain
        </label>
        <input
          class="chakra-input css-qcorp"
          data-testid="terrain"
          id="terrain"
          name="terrain"
          placeholder="Terrain"
          value=""
        />
      </div>
      <div
        class="css-u8chpe"
      >
        <button
          class="chakra-button css-4g9tt9"
          type="button"
        >
          Cancel
        </button>
        <button
          class="chakra-button css-12euwm4"
          disabled=""
          type="submit"
        >
          Create
        </button>
      </div>
    </form>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
